### PR TITLE
Added wise_enum header only recipe

### LINKED ISSUE #7172: [request] <wise_enum>/3.0.0
### Package Details
  * Package Name/Version: **wise_enum/3.0.0**
  * Website: **https://github.com/quicknir/wise_enum#readme**
  * Source code: **https://github.com/quicknir/wise_enum*

### Description Of The Library / Tool
A small header only c++ library for reflective enum. simular to magic enum, but supports C++ 14/11

### PR DESCRIPTION
**Pull request by szavadsky**
Closes #7172
Package Details
Package Name/Version: wise_enum/3.0.0
Website: https://github.com/quicknir/wise_enum#readme
Source code: *https://github.com/quicknir/wise_enum
Description Of The Library / Tool
A small header-only c++ library for reflective enum. similar to magic enum, but supports C++ 14/11
I am not the author of the library, but I would like to use it for the project that already uses conan

---

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [x] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [x] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/wise_enum/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
+from conans import ConanFile, tools
+from conans.errors import ConanInvalidConfiguration
+import os
+
+class WiseEnumConan(ConanFile):
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:5]: Needed for my suggestion in the source method.
```suggestion
import os

required_conan_version = ">=1.33.0"

class WiseEnumConan(ConanFile):
```

File: recipes/wise_enum/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    name = "wise_enum"
+    description = "Header-only C++11/14/17 library provides static reflection for enums, work with any enum type without any boilerplate code."
+    topics = (
+        "conan",
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:9]: We have stopped adding `conan` to topics.
```suggestion
```

File: recipes/wise_enum/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+        "cplusplus",
+        "enum-to-string",
+        "string-to-enum"
+        "serialization",
+        "reflection",
+        "header-only",
+        "compile-time"
+    )
+    url = "https://github.com/quicknir/wise_enum"
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:18]:
```suggestion
    homepage = "https://github.com/quicknir/wise_enum"
    url = "https://github.com/conan-io/conan-center-index/"
```

File: recipes/wise_enum/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    @property
+    def _source_subfolder(self):
+        return "src"
+    
+   
+
+    def configure(self):
+        if self.settings.compiler.get_safe("cppstd"):
+            tools.check_min_cppstd(self, "11")
+    def package_id(self):
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:33]:
```suggestion
    no_copy_source = True
    
    @property
    def _source_subfolder(self):
        return "source_subfolder"    
   

    def configure(self):
        if self.settings.compiler.get_safe("cppstd"):
            tools.check_min_cppstd(self, "11")

    def package_id(self):
```

File: recipes/wise_enum/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        self.info.header_only()
+
+    def source(self):
+        tools.get(**self.conan_data["sources"][self.version])
+        extracted_dir = self.name + "-" + self.version
+        os.rename(extracted_dir, self._source_subfolder)
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:39]:
```suggestion
        tools.get(**self.conan_data["sources"][self.version],
                  destination=self._source_subfolder, strip_root=True)
```

File: recipes/wise_enum/all/conanfile.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+    def package(self):
+        self.copy("*.h", dst="include", src=self._source_subfolder)
+        self.copy("LICENSE", dst="licenses" , src=self._source_subfolder)
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:43]: We explicitly set the names for the generator when the upstream project creates one.
https://github.com/quicknir/wise_enum/blob/34ac79f7ea2658a148359ce82508cc9301e31dd3/CMakeLists.txt#L66
```suggestion
        self.copy("LICENSE", dst="licenses" , src=self._source_subfolder)
        
    def package_info(self):
        self.cpp_info.names["cmake_find_package"] = "wise_enum"
        self.cpp_info.names["cmake_find_package_multi"] = "wise_enum"
```

File: recipes/wise_enum/all/test_package/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.8)
+project(test_package)
```
Comment (madebr) [inline recipes/wise_enum/all/test_package/CMakeLists.txt:2]:
```suggestion
project(test_package CXX)
```

File: recipes/wise_enum/all/test_package/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
...
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake", "cmake_find_package_multi"
+
+    def build(self):
+        cmake = CMake(self)
+        cmake.configure()
+        cmake.build()
+
+    def test(self):
+        if not tools.cross_building(self.settings):
```
Comment (madebr) [inline recipes/wise_enum/all/test_package/conanfile.py:15]:
```suggestion
        if not tools.cross_building(self):
```

File: recipes/wise_enum/all/conanfile.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+        "enum-to-string",
+        "string-to-enum"
+        "serialization",
+        "reflection",
+        "header-only",
+        "compile-time"
+    )
+    homepage = "https://github.com/quicknir/wise_enum"
+    url = "https://github.com/conan-io/conan-center-index/"
+    homepage = url
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:21]:
```suggestion
```

File: recipes/wise_enum/all/conanfile.py [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    license = "BSL-1.0"
+    settings = "compiler"
+    no_copy_source = True
+    
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"    
+   
+
+    def configure(self):
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:31]:
```suggestion
    def validate(self):
```

File: recipes/wise_enum/all/conanfile.py [hunk 11]
```diff
@@ -0,0 +1,48 @@
...
+        "cplusplus",
+        "enum-to-string",
+        "string-to-enum"
+        "serialization",
+        "reflection",
+        "header-only",
+        "compile-time"
+    )
+    homepage = "https://github.com/quicknir/wise_enum"
+    url = "https://github.com/conan-io/conan-center-index/"
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:20]: It looks like other recipes don't have a trailing `/`.
```suggestion
    url = "https://github.com/conan-io/conan-center-index"
```

File: recipes/wise_enum/all/conanfile.py [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        self.cpp_info.names["cmake_find_package"] = "wise_enum"
+        self.cpp_info.names["cmake_find_package_multi"] = "wise_enum"
+        
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "wise_enum"
+        self.cpp_info.names["cmake_find_package_multi"] = "wise_enum"
+        
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "wise_enum"
+        self.cpp_info.names["cmake_find_package_multi"] = "wise_enum"
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:55]: I don't know how you did it, but you applied my suggestion 3 times :)
```suggestion
        self.cpp_info.names["cmake_find_package_multi"] = "wise_enum"
```

File: recipes/wise_enum/all/conanfile.py [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:59]: I don't know how you did it, but you applied my suggestion 4 times.
```suggestion
        self.cpp_info.names["cmake_find_package_multi"] = "wise_enum"
```

Comment (szavadsky) [reply recipes/wise_enum/all/conanfile.py]: Yes, side effect of pressing on my phone.
Unapplied
Thx for review btw, I learned a lot

File: recipes/wise_enum/all/conanfile.py [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+    )
+    homepage = "https://github.com/quicknir/wise_enum"
+    url = "https://github.com/conan-io/conan-center-index"
+    license = "BSL-1.0"
+    settings = "compiler"
+    no_copy_source = True
+    
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
```
Comment (uilianries) [inline recipes/wise_enum/all/conanfile.py:28]:
```suggestion
        return "source_subfolder"

```

File: recipes/wise_enum/all/conanfile.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        compiler = str(self.settings.compiler)
+        compiler_version = tools.Version(self.settings.compiler.version)
+
+        minimal_version = {
+           "gcc": "5"
+        }
+        unsupported = {"Visual Studio"}
+        if compiler in unsupported:
+            raise ConanInvalidConfiguration(
+                "{} doe not support  {} compiler".format(self.name, compiler)
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:53]:
```suggestion
                "{} does not support  {} compiler".format(self.name, compiler)
```

File: recipes/wise_enum/all/conanfile.py [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+    def _python_executable(self):
+        """
+        obtain full path to the python interpreter executable
+        :return: path to the python interpreter executable, either set by option, or system default
+        """
+        exe = sys.executable
+        return str(exe).replace('\\', '/')  
+   
+
+    def validate(self):
```
Comment (madebr) [inline recipes/wise_enum/all/conanfile.py:40]: Not used.
```suggestion
        return "source_subfolder"

    def validate(self):
```

**Review by madebr — APPROVED** [review]

**Review by uilianries — DISMISSED** [review]

File: recipes/wise_enum/all/conanfile.py [hunk 17]
```diff
@@ -0,0 +1,66 @@
...
+                  destination=self._source_subfolder, strip_root=True)
+        
+    def package(self):
+        self.copy("*.h", dst="include", src=self._source_subfolder)
+        self.copy("LICENSE", dst="licenses" , src=self._source_subfolder)
+        
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "wise_enum"
+        self.cpp_info.names["cmake_find_package_multi"] = "wise_enum"
+
```
Comment (SpaceIm) [inline recipes/wise_enum/all/conanfile.py:66]:
```suggestion
        self.cpp_info.names["cmake_find_package"] = "WiseEnum"
        self.cpp_info.names["cmake_find_package_multi"] = "WiseEnum"
        self.cpp_info.names["pkg_config"] = "WiseEnum"
        self.cpp_info.components["_wise_enum"].names["cmake_find_package"] = "wise_enum"
        self.cpp_info.components["_wise_enum"].names["cmake_find_package"] = "wise_enum"
        self.cpp_info.components["_wise_enum"].names["pkg_config"] = "WiseEnum"
```

File: recipes/wise_enum/all/test_package/CMakeLists.txt [hunk 18]
```diff
@@ -0,0 +1,11 @@
+cmake_minimum_required(VERSION 3.8)
+project(test_package CXX)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup(TARGETS)
+
+find_package(wise_enum REQUIRED CONFIG)
```
Comment (SpaceIm) [inline recipes/wise_enum/all/test_package/CMakeLists.txt:7]:
```suggestion
find_package(WiseEnum REQUIRED CONFIG)
```

File: recipes/wise_enum/all/test_package/CMakeLists.txt [hunk 19]
```diff
@@ -0,0 +1,11 @@
...
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+set_property(TARGET ${PROJECT_NAME} PROPERTY CXX_STANDARD 11)
+target_link_libraries(${PROJECT_NAME} wise_enum::wise_enum)
```
Comment (SpaceIm) [inline recipes/wise_enum/all/test_package/CMakeLists.txt:11]:
```suggestion
target_link_libraries(${PROJECT_NAME} WiseEnum::wise_enum)
```

**Review by SpaceIm — CHANGES_REQUESTED** [review]

File: recipes/wise_enum/all/conanfile.py [hunk 20]
```diff
@@ -0,0 +1,69 @@
...
+    def package(self):
+        self.copy("*.h", dst="include", src=self._source_subfolder)
+        self.copy("LICENSE", dst="licenses" , src=self._source_subfolder)
+        
+    def package_info(self):
+        self.cpp_info.names["cmake_find_package"] = "WiseEnum"
+        self.cpp_info.names["cmake_find_package_multi"] = "WiseEnum"
+        self.cpp_info.names["pkg_config"] = "WiseEnum"
+        self.cpp_info.components["_wise_enum"].names["cmake_find_package"] = "wise_enum"
+        self.cpp_info.components["_wise_enum"].names["cmake_find_package"] = "wise_enum"
```
Comment (SpaceIm) [inline recipes/wise_enum/all/conanfile.py:68]:
```suggestion
        self.cpp_info.components["_wise_enum"].names["cmake_find_package_multi"] = "wise_enum"
```

sorry, typo

**Review by SpaceIm — APPROVED** [review]

**Review by madebr — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Pull request merged**